library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt

```diff
@@ -1056,86 +1202,118 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
         }
         mMinZoom = minZoom
         mMinZoomMode = type
-        if (zoom <= resolveZoom(minZoom, type)) {
-            zoomTo(resolveZoom(minZoom, type), true)
+        if (zoom <= getMinZoom()) {
+            zoomTo(getMinZoom(), animate = true)
         }
     }
 
     //endregion
 
     //region Apply values
 
+    private val mCancelAnimationListener = object : Animator.AnimatorListener {
+        override fun onAnimationStart(animation: Animator?) {
+        }
+
+        override fun onAnimationEnd(animation: Animator?) {
+            setState(NONE)
+        }
+
+        override fun onAnimationCancel(animation: Animator?) {
+        }
+
+        override fun onAnimationRepeat(animation: Animator?) {
+        }
+    }
+
+    /**
+     * Prepares a [ValueAnimator] for the first run
+     */
+    private fun ValueAnimator.prepare() {
+        this.duration = mAnimationDuration
+        this.addListener(mCancelAnimationListener)
+        this.interpolator = ANIMATION_INTERPOLATOR
+    }
+
     /**
      * Calls [applyZoom] repeatedly
      * until the final zoom is reached, interpolating.
      *
-     * @param newZoom        the new zoom
+     * @param zoom        the new zoom
      * @param allowOverPinch whether overpinching is allowed
      */
-    private fun animateZoom(@Zoom newZoom: Float, allowOverPinch: Boolean) {
-        var newZoom = newZoom
-        newZoom = checkZoomBounds(newZoom, allowOverPinch)
+    private fun animateZoom(@Zoom zoom: Float, allowOverPinch: Boolean) {
         if (setState(ANIMATING)) {
             mClearAnimation = false
-            val startTime = System.currentTimeMillis()
-            @Zoom val startZoom = zoom
-            @Zoom val endZoom = newZoom
-            mContainer.post(object : Runnable {
-                override fun run() {
-                    if (mClearAnimation) return
-                    val time = interpolateAnimationTime(System.currentTimeMillis() - startTime)
-                    LOG.v("animateZoomAndAbsolutePan:", "animationStep:", time)
-                    @Zoom val zoom = startZoom + time * (endZoom - startZoom)
-                    applyZoom(zoom, allowOverPinch)
-                    if (time >= 1f) {
-                        setState(NONE)
-                    } else {
-                        mContainer.postOnAnimation(this)
-                    }
+            @Zoom val startZoom = this.zoom
+            @Zoom val endZoom = checkZoomBounds(zoom, allowOverPinch)
+            val zoomAnimator = ValueAnimator.ofFloat(startZoom, endZoom)
+            zoomAnimator.prepare()
+            zoomAnimator.addUpdateListener {
+                LOG.v("animateZoom:", "animationStep:", it.animatedFraction)
+                if (mClearAnimation) {
+                    it.cancel()
                 }
-            })
+                mContainer.postOnAnimation {
```

Comment:
You don't need this post here, I think the animator already runs in an animation step.